From f9e98fe51a62d9eef48b70f1b12f236627e7fa85 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:38:47 -0400
Subject: [alsa] pcm: Add timestamp type to sw_params

Message-id: <140972823573-387-patchset-jkysela@redhat.com>
Patchwork-id: 91083
O-Subject: [RHEL 7.1 PATCH 386/442] ALSA: pcm: Add timestamp type to sw_params
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

For allowing adjusting the timestamp type on the fly, add it to
sw_params.  The existing ioctl is still kept for compatibility.

Along with this, increment the PCM protocol version.

The extension was suggested by Clemens Ladisch.

Acked-by: Jaroslav Kysela <perex@perex.cz>
Reviewed-by: Mark Brown <broonie@linaro.org>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: 5646eda5851e6cfdfa22d41895e3f5daffa643d3)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/uapi/sound/asound.h b/include/uapi/sound/asound.h
index 041203f..3204471 100644
--- a/include/uapi/sound/asound.h
+++ b/include/uapi/sound/asound.h
@@ -136,7 +136,7 @@ struct snd_hwdep_dsp_image {
  *                                                                           *
  *****************************************************************************/
 
-#define SNDRV_PCM_VERSION  SNDRV_PROTOCOL_VERSION(2, 0, 11)
+#define SNDRV_PCM_VERSION  SNDRV_PROTOCOL_VERSION(2, 0, 12)
 
 typedef unsigned long snd_pcm_uframes_t;
 typedef signed long snd_pcm_sframes_t;
@@ -388,7 +388,9 @@ struct snd_pcm_sw_params {
  snd_pcm_uframes_t silence_threshold; /* min distance from noise for silence filling */
  snd_pcm_uframes_t silence_size;  /* silence block size */
  snd_pcm_uframes_t boundary;  /* pointers wrap point */
- unsigned char reserved[64];  /* reserved for future */
+ unsigned int tstamp_type;  /* timestamp type */
+ int pads;    /* alignment, reserved */
+ unsigned char reserved[56];  /* reserved for future */
 };
 
 struct snd_pcm_channel_info {
diff --git a/sound/core/pcm_native.c b/sound/core/pcm_native.c
index d1afafe..27de230 100644
--- a/sound/core/pcm_native.c
+++ b/sound/core/pcm_native.c
@@ -543,6 +543,8 @@ static int snd_pcm_sw_params(struct snd_pcm_substream *substream,
 
  if (params->tstamp_mode > SNDRV_PCM_TSTAMP_LAST)
   return -EINVAL;
+ if (params->tstamp_type > SNDRV_PCM_TSTAMP_TYPE_LAST)
+  return -EINVAL;
  if (params->avail_min == 0)
   return -EINVAL;
  if (params->silence_size >= runtime->boundary) {
@@ -557,6 +559,7 @@ static int snd_pcm_sw_params(struct snd_pcm_substream *substream,
  err = 0;
  snd_pcm_stream_lock_irq(substream);
  runtime->tstamp_mode = params->tstamp_mode;
+ runtime->tstamp_type = params->tstamp_type;
  runtime->period_step = params->period_step;
  runtime->control->avail_min = params->avail_min;
  runtime->start_threshold = params->start_threshold;
-- 
1.7.1